### PR TITLE
feat: Auto merge and deploy to staging branch

### PR DESCRIPTION
**Pull request by prathapsridharan**
## Reason for Change

- #5461

The idea is that we can leverage our existing workflow where a commit to `staging` branch creates a deployment to the `stage` environment. So, whenever new commits to `main` come in, we merge the commit to `staging` thus kicking off a deployment to `stage` environment.

Additionally, github action's [concurrency group feature is utilized](https://docs.github.com/en/actions/using-jobs/using-concurrency) to eliminate race conditions involving multiple builds and deployments run at the same time. Here, we cancel scheduled builds and deployments as well as currently running builds and deployments in favor of the newest build/deployment job.

## Changes

- add merging `main` branch to `staging` branch on new commits to `main`
- add GHA concurrency group feature

## Testing steps

- The following workflow was used to test a merge between 2 branches not named `[dev, staging, prod]`. Specifically the following github workflow merges commits from `psridharan/issue-5461` into `psridharan/test-auto-merge-2` triggered by a pushes to `psridharan/issue-5461`:

```
name: Trigger release candidate build and deploy

on:
  push:
    branches:
      - psridharan/issue-5461

jobs:
  merge:
    runs-on: ubuntu-latest
    steps:
      - name: Checkout psridharan/issue-5461 into psridharan/test-auto-merge-2
        uses: actions/checkout@v2
        with:
          ref: psridharan/issue-5461
          fetch-depth: 0
      - name: Merge psridharan/issue-5461 into psridharan/test-auto-merge-2
        run: |
          git config user.name "GitHub Actions"
          git config user.email "actions@github.com"

          echo "BEFORE CHECKOUT: Listing current checked out branches:"
          echo "$(git branch --list)"

          echo "Checking out psridharan/test-auto-merge-2"
          git checkout psridharan/test-auto-merge-2
          echo "Confirming checked out branch receiving merge is: $(git branch --show-current)"

          echo "AFTER CHECKOUT: Listing current checked out branches:"
          echo "$(git branch --list)"

          echo "Current checked out receiving branch is: $(git branch --show-current)"
          echo "HEAD commit in current checked out receiving branch: $(git branch --show-current) is: $(git log -1 --format='%H')"

          echo "HEAD commit in incoming branch psridharan/issue-5461 is:"
          echo "$(git log -1 --format='%H' psridharan/issue-5461)"

          echo "About to merge psridharan/issue-5461 into psridharan/test-auto-merge-2"
          git merge --verbose --no-ff psridharan/issue-5461 -m "Merging psridharan/issue-5461 into psridharan/test-auto-merge-2"

          git push origin psridharan/test-auto-merge-2
```

- Here is the DEBUG log output that inspects the state before the `git merge` operation:

```
BEFORE CHECKOUT: Listing current checked out branches:
* psridharan/issue-5461
Checking out psridharan/test-auto-merge-2
Switched to a new branch 'psridharan/test-auto-merge-2'
branch 'psridharan/test-auto-merge-2' set up to track 'origin/psridharan/test-auto-merge-2'.
Confirming checked out branch receiving merge is: psridharan/test-auto-merge-2
AFTER CHECKOUT: Listing current checked out branches:
  psridharan/issue-5461
* psridharan/test-auto-merge-2
Current checked out receiving branch is: psridharan/test-auto-merge-2
HEAD commit in current checked out receiving branch: psridharan/test-auto-merge-2 is: baf69efb5f6f94c2d6645e3eedaf4d8703852ccc
HEAD commit in incoming branch psridharan/issue-5461 is:
4b12fc3e[27](https://github.com/chanzuckerberg/single-cell-data-portal/actions/runs/6386570986/job/17333448676?pr=5860#step:3:28)50f6af67bae5ca[28](https://github.com/chanzuckerberg/single-cell-data-portal/actions/runs/6386570986/job/17333448676?pr=5860#step:3:29)354502e9e[33](https://github.com/chanzuckerberg/single-cell-data-portal/actions/runs/6386570986/job/17333448676?pr=5860#step:3:34)bba
About to merge psridharan/issue-5461 into psridharan/test-auto-merge-2
Merge made by the 'ort' strategy.
 .../build-images-and-create-deployment.yml         |  4 +++
 .github/workflows/deploy-happy-stack.yml           |  5 +++
 .../trigger-release-candidate-build-and-deploy.yml | 41 ++++++++++++++++++++++
 3 files changed, 50 insertions(+)
 create mode 100644 .github/workflows/trigger-release-candidate-build-and-deploy.yml
To https://github.com/chanzuckerberg/single-cell-data-portal
   baf69efb5f6..64a[37](https://github.com/chanzuckerberg/single-cell-data-portal/actions/runs/6386570986/job/17333448676?pr=5860#step:3:38)11d333  psridharan/test-auto-merge-2 -> psridharan/test-auto-merge-2
```

Attached is an image that shows the auto-merge

Screenshot 2023-10-02 at 4 31 47 PM

### CODE REVIEW

File: .github/workflows/merge-main-into-staging.yml
Comment (atolopko-czi) [inline .github/workflows/merge-main-into-staging.yml:14]: Might rename this to indicate the underlying intention: "Trigger release candidate build and deploy". In fact maybe rename the entire workflow similarly?

File: .github/workflows/merge-main-into-staging.yml
Comment (atolopko-czi) [inline .github/workflows/merge-main-into-staging.yml:13]: consider using `ref: staging` instead of `git checkout staging` later.

We might also be able to get away with `fetch-depth: 1` (for checkout speed), if we later do a merge from main as `git pull origin main`

File: .github/workflows/merge-main-into-staging.yml
Comment (atolopko-czi) [inline .github/workflows/merge-main-into-staging.yml:20]: Is the `--no-ff` strictly necessary? Just curious.

**Review by atolopko-czi — CHANGES_REQUESTED** [review]
Exciting! A couple suggestions for your consideration.

File: .github/workflows/merge-main-into-staging.yml
Comment (prathapsridharan) [reply .github/workflows/merge-main-into-staging.yml:20]: _With the `--no-ff` option on it creates a merge commit in all cases_. This means every merge into `staging` will be merge commit even if the current HEAD of `staging` is an ancestor of the incoming commit from `main`. Without the `--no-ff` you won't see a merge commit on `staging` for the above case - instead the branch pointer would be moved up to point to the commit at the HEAD of `main`.

From the perspective of someone looking at the `staging` branch, seeing a merge commit like `Merge ...` more easily allows them to see that the GHA executed and what was added from `main` when this GHA executes.

I could remove it if you feel strongly that explicit merge commits are not helpful and are too noisy

File: .github/workflows/merge-main-into-staging.yml
Comment (prathapsridharan) [reply .github/workflows/merge-main-into-staging.yml]: will fix

File: .github/workflows/merge-main-into-staging.yml
Comment (prathapsridharan) [reply .github/workflows/merge-main-into-staging.yml]: will use `ref: staging`

File: .github/workflows/trigger-release-candidate-build-and-deploy.yml
Comment (atarashansky) [inline .github/workflows/trigger-release-candidate-build-and-deploy.yml:16]: We set fetch-depth to 1 in order to prevent the checkout step from taking upwards of 10 minutes (our repo is super bloated and fetching all the commits takes forever). Is it possible to figure out how to fetch only the commits that are needed?

**Review by atarashansky — CHANGES_REQUESTED** [review]

File: .github/workflows/trigger-release-candidate-build-and-deploy.yml
Comment (prathapsridharan) [reply .github/workflows/trigger-release-candidate-build-and-deploy.yml:16]: @atarashansky - I _think_ it is possible. Perhaps there is a way to use `git fetch --shallow-since`, but I am not sure yet.

But since this work is towards an MVP of continuous delivery to `staging`, I don't think this PR needs to be blocked on the fetch-depth issue. If it is taking too long, we can always turn off this workflow.

There are a couple of challenges with figuring out the issue of fetching only the commits from the incoming `main` branch that is:

1. Testing is pretty onerous for GHA stuff. As far as my knowledge is goes with GHA, I would have to keep pushing to this PR to demonstrate a test with test branches. Just look at the number of commits I had to make just to demonstrate that correctness of auto-merge

2. Concurrency poses problems. Say you have multiple commits into `main` in rapid succession that trigger this merge to `staging`. From my reading, there is no order guarantee of how these commits are processed. I don't think something like `fetch-depth=x` will work in this case. The fetching scheme under concurrency needs to be tested for correctness and the more complex the fetching scheme is, the harder it is to test, and the less sure we are about its correctness under concurrency. `fetch-depth=0` does not have this problem under concurrency.

My suggestion is to tackle the fetching scheme in a subsequent PR rather than hold this back for an MVP.

Comment (atarashansky) [reply .github/workflows/trigger-release-candidate-build-and-deploy.yml:16]: OK - sounds good!

**Review by atarashansky — APPROVED** [review]

File: .github/workflows/build-images-and-create-deployment.yml
Comment (Bento007) [inline .github/workflows/build-images-and-create-deployment.yml:11]: Is this a new feature? I love it! We should add something like this for rdev created for a PR.

File: .github/workflows/trigger-release-candidate-build-and-deploy.yml
Comment (Bento007) [inline .github/workflows/trigger-release-candidate-build-and-deploy.yml:17]: I thought the plan was deploy main to the staging environment and removing the staging branch?

File: .github/workflows/trigger-release-candidate-build-and-deploy.yml
Comment (Bento007) [inline .github/workflows/trigger-release-candidate-build-and-deploy.yml:22]: Why do we need to do this list?

File: .github/workflows/trigger-release-candidate-build-and-deploy.yml
Comment (Bento007) [inline .github/workflows/trigger-release-candidate-build-and-deploy.yml:32]: > "HEAD commit in current checked out receiving branch:

I'm not sure what this means?

**Review by Bento007 — COMMENTED** [review]
Some of the language you use in the echo statement is confusing to understand. Is there anyway to make it more clear.

**Comment by prathapsridharan** [comment]
> Some of the language you use in the echo statement is confusing to understand. Is there anyway to make it more clear.

@Bento007 - I can clarify some of the echo messages. I initially had them for debugging to understand the state prior to merge but I will clean them up

File: .github/workflows/trigger-release-candidate-build-and-deploy.yml
Comment (prathapsridharan) [reply .github/workflows/trigger-release-candidate-build-and-deploy.yml:32]: will clean the echo messages up

File: .github/workflows/trigger-release-candidate-build-and-deploy.yml
Comment (prathapsridharan) [reply .github/workflows/trigger-release-candidate-build-and-deploy.yml:17]: @Bento007 (cc: @atolopko-czi ) -

Here we want to do the simplest thing to demonstrate the pipeline end-to-end. It is not the most efficient, but deploying `main` to `staging` and reusing the existing `deploy-happy-stack.yml` to deploy to `stage` env seems to be the simplest thing to do.

The most efficient route is to build only once and deploy the builds to different environments (`staging`, `prod`, etc) but that is a bit complex in terms of implementation and testing.

There are some additional benefits to this simple approach too:

1. Easier to test. GHA, as far as I understand, is pretty onerous to test. The number of commits in this PR to demonstrate auto-merge is an example of the pain point. The simplest path to the solution also happens to be the shortest path - just merge `main` to `staging` - which means easier to demonstrate a test of correctness.
2. Concurrency introduces complexity (ex: many commits to `main` coming in at once) and to the extent we can keep things straightforward, will allow us to understand the complexities of concurrency and then we can determine a more efficient solution
3. Easier to reason about in the beginning (Uniformity) - The merge-commit that goes from `main` to `staging` is replicates our documented manual steps for a `stage` deploy thus making continuous deployment to staging easier to reason about.
4. Easy to rollback. If continuous deployment to `stage` is running into problems, it is as simple as disabling the GHA workflow that merges `main` to `staging` and we are back to our starting point.

File: .github/workflows/build-images-and-create-deployment.yml
Comment (prathapsridharan) [reply .github/workflows/build-images-and-create-deployment.yml:11]: @Bento007 - I don't know if it is a new feature but this was suggested by @atolopko-czi when I brought up the issue of concurrency and order of operations:

https://github.com/chanzuckerberg/single-cell-data-portal/issues/5461#issuecomment-1740899948

**Review by Bento007 — APPROVED** [review]

File: .github/workflows/build-images-and-create-deployment.yml
Comment (atolopko-czi) [inline .github/workflows/build-images-and-create-deployment.yml:13]: curious, what is the issue with canceling "in progress" workflows?

File: .github/workflows/trigger-release-candidate-build-and-deploy.yml
Comment (atolopko-czi) [reply .github/workflows/trigger-release-candidate-build-and-deploy.yml:16]: agreed, reasonble to take on later. can you add an issue to the [post-mvp](https://app.zenhub.com/workspaces/single-cell-5e2a191dad828d52cc78b028/issues/gh/chanzuckerberg/single-cell-data-portal/5457) epic?

Comment (atolopko-czi) [reply .github/workflows/trigger-release-candidate-build-and-deploy.yml:16]: some possible solutions usings --merge-base and --deepen options, referenced in these [search results](https://www.google.com/search?q=git+clone+sufficient+history+to+merge+two+branches), but no clear immediate win.

File: .github/workflows/trigger-release-candidate-build-and-deploy.yml
Comment (atolopko-czi) [reply .github/workflows/trigger-release-candidate-build-and-deploy.yml:17]: I agree w/most of this and the overall decision.

For 2, haven't we addressed the concurrency issue with `concurrency` config? If not, isn't the concurrency concern for `staging` branch merges going to be the same as it would be for `main`-based staging deploys?

For 1, most of the GHA-dev/debug cycle was for the main->staging auto-merge right? So possibly might have been shorted dev time if deployed to stagin from `main`? (I haven't studies all the commits!)

File: .github/workflows/build-images-and-create-deployment.yml
Comment (atolopko-czi) [inline .github/workflows/build-images-and-create-deployment.yml:1]: Shouldn't we also remove `main` from `on: push: branches:` now?

File: .github/workflows/trigger-release-candidate-build-and-deploy.yml
Comment (atolopko-czi) [reply .github/workflows/trigger-release-candidate-build-and-deploy.yml:16]: Another way entirely around merging would be to destroy and recreate the staging branch every time.

File: .github/workflows/build-images-and-create-deployment.yml
Comment (Bento007) [reply .github/workflows/build-images-and-create-deployment.yml:13]: I don't know the exact reason but there is a lot of state changes that happens as a result of the deployment. At the very least there are test fixtures created and happy deployment configuration files that are modified. Interrupting a deployment could leave it in an unknown state. If we do cancel in progress workflows we should be mindful of when and where we allow it. I think determining that would be outside the scope of this PR.

File: .github/workflows/trigger-release-candidate-build-and-deploy.yml
Comment (prathapsridharan) [reply .github/workflows/trigger-release-candidate-build-and-deploy.yml:17]: @atolopko-czi 

For concurrency (2), we have a provisional solution with `concurrency` feature but we haven't tested it. My point is that it is hard to see right now what kinds of concurrency gotchas we might run into.

For the testing point made in (1), if I understand you correctly, you are proposing committing the change to `main` and then see if it actually auto-merges to `staging`? In that case it would take the same number of attempts (see the number of commits to test auto-merge between test branches) and moreover, pollute `main` with a bunch of these "attempting-fix" commits.

In the absence of unit tests, I think it is preferable to demonstrate tests in testing sandbox - in this case, using a couple of test branches

File: .github/workflows/build-images-and-create-deployment.yml
Comment (prathapsridharan) [reply .github/workflows/build-images-and-create-deployment.yml:13]: @atolopko-czi @Bento007 - `cancel-in-progress: true` seems like a blunt force instrument. It might open the [possibility of corrupting state](https://github.com/orgs/community/discussions/26311). I think it `cancel-in-progress: true` could work out for us, but I feel starting conservatively here is the best way to go.

Note that not setting `cancel-in-progress` (`false` is the default), won't prevent the `concurrency` feature from [canceling pending tasks](https://docs.github.com/en/actions/using-jobs/using-concurrency)

So we still save on time here by using the default `concurrency` feature

File: .github/workflows/build-images-and-create-deployment.yml
Comment (prathapsridharan) [reply .github/workflows/build-images-and-create-deployment.yml:1]: We still want a `dev` deployment right as per the ticket: https://github.com/chanzuckerberg/single-cell-data-portal/issues/5461

Removing `dev` deployment is in [another ticket](https://github.com/chanzuckerberg/single-cell-data-portal/issues/5847) and that is when we remove `main` from `on: push: branches:` in this file I think

Comment (atolopko-czi) [reply .github/workflows/build-images-and-create-deployment.yml:1]: 💯 Forgot about the issue split!

File: .github/workflows/build-images-and-create-deployment.yml
Comment (atolopko-czi) [reply .github/workflows/build-images-and-create-deployment.yml:13]: Thanks, both. Good justifications. Maybe just add a TODO and/or issue to revisit this later?

File: .github/workflows/trigger-release-candidate-build-and-deploy.yml
Comment (atolopko-czi) [reply .github/workflows/trigger-release-candidate-build-and-deploy.yml:17]: > For the testing point made in (1), if I understand you correctly, you are proposing committing the change to main and then see if it actually auto-merges to staging?  

Not actually proposing anything :) We can take this offline, since I was just thinking aloud.

File: .github/workflows/build-images-and-create-deployment.yml
Comment (prathapsridharan) [reply .github/workflows/build-images-and-create-deployment.yml:13]: @atolopko-czi - Added a ticket here: https://github.com/chanzuckerberg/single-cell-data-portal/issues/5895

**Review by atolopko-czi — APPROVED** [review]
LGTM, with 2 requests to add follow-up issues for future improvements:
- remove need for fetch-depth=0 (full clone)
- revisit canceling in-progress builds

File: .github/workflows/trigger-release-candidate-build-and-deploy.yml
Comment (prathapsridharan) [reply .github/workflows/trigger-release-candidate-build-and-deploy.yml:16]: @atolopko-czi - I have created a ticket here to capture the fact that we might want to improve the speed of a `stage` deploy: https://github.com/chanzuckerberg/single-cell-data-portal/issues/5896

**Comment by prathapsridharan** [comment]
> LGTM, with 2 requests to add follow-up issues for future improvements:
> 
> * remove need for fetch-depth=0 (full clone)
> * revisit canceling in-progress builds

@atolopko-czi  - Added issues #5895 and #5896 to address these

File: .github/workflows/merge-main-into-staging.yml
Comment (danieljhegeman) [reply .github/workflows/merge-main-into-staging.yml]: i might argue against it since we will be deploying ~continuously~ 😄

Comment (atolopko-czi) [reply .github/workflows/merge-main-into-staging.yml]: Agreed--since _every commit_ to main is going to be merged to staging, it's probably not adding much to see the merge commit.

Comment (prathapsridharan) [reply .github/workflows/merge-main-into-staging.yml]: will remove

**Pull request merged**
